src/java.base/share/classes/java/math/MutableBigInteger.java

```diff
@@ -1882,6 +1954,201 @@ MutableBigInteger sqrt() {
+             * for a long value in the range [0, 2^64) is either correct,
+             * or rounded up by one if the value is too high
+             * and too close to the next perfect square.
+             */
+            long s = Math.round(Math.sqrt(x >= 0 ? x : x + 0x1p64));
+            if (s > LONG_MASK // avoid overflow of s * s
+                    || Long.compareUnsigned(x, s * s) < 0) {
+                s--;
+            }
+
+            return new MutableBigInteger[] {
+                    new MutableBigInteger((int) s),
+                    needRemainder ? new MutableBigInteger(x - s * s) : null
+            };
+        }
+
+        long bitLength = this.bitLength();
+        if (bitLength != (int) bitLength)
+            throw new ArithmeticException("bitLength() integer overflow");
+
+        // Normalize
+        MutableBigInteger x = this;
+        int shift = Integer.numberOfLeadingZeros(x.value[x.offset]) & ~1; // shift must be even
+        if (shift != 0) {
+            x = new MutableBigInteger(x);
+            x.primitiveLeftShift(shift);
+        }
+
+        MutableBigInteger[] sqrtRem = x.sqrtRemZimmermann(x.intLen, needRemainder);
+
+        // Unnormalize
+        if (shift != 0) {
+            final int halfShift = shift >> 1;
+            if (needRemainder && !sqrtRem[1].isZero()) {
+                final int s0 = sqrtRem[0].value[sqrtRem[0].offset + sqrtRem[0].intLen - 1]
+                        & (-1 >>> -halfShift); // Remove excess bits
+                if (s0 != 0) { // An optimization
+                    // s0 is at most 15 bit long
+                    MutableBigInteger doubleProd = new MutableBigInteger();
+                    sqrtRem[0].mul(s0 << 1, doubleProd);
+
+                    sqrtRem[1].add(doubleProd);
+                    sqrtRem[1].subtract(new MutableBigInteger(s0 * s0));
+                }
+                sqrtRem[1].primitiveRightShift(shift);
+            }
+            sqrtRem[0].primitiveRightShift(halfShift);
+        }
+        return sqrtRem;
+    }
+
+    /**
+     * Assume {@code 2 <= len <= intLen && Integer.numberOfLeadingZeros(value[offset]) <= 1}
+     * @implNote The implementation is based on Zimmermann's works available
+     * <a href="https://inria.hal.science/inria-00072854/en/">  here</a> and
+     * <a href="https://www.researchgate.net/publication/220532560_A_proof_of_GMP_square_root">  here</a>
+     */
+    private MutableBigInteger[] sqrtRemZimmermann(int len, boolean needRemainder) {
+        if (len == 2) { // Base case
+            long x = ((value[offset] & LONG_MASK) << 32) | (value[offset + 1] & LONG_MASK);
+            long s = Math.round(Math.sqrt(x >= 0 ? x : x + 0x1p64));
+            if (s > LONG_MASK || Long.compareUnsigned(x, s * s) < 0)
+                s--;
+
+            // Allocate sufficient space to hold the normalized final square root
+            MutableBigInteger sqrt = new MutableBigInteger(new int[(intLen + ((-intLen) & 3)) >> 1]);
```

Comment:
Rather than expecting the reader to stop for some minutes to figure out what's happening with the arithmetic here, add a comment or prefer
```suggestion
            MutableBigInteger sqrt = new MutableBigInteger(new int[2 * Math.ceilDiv(intLen, 4)]);
```